File: Data Structures & Algorithms/linked-list-cycle-detection/submission-2.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

func hasCycle(head *ListNode) bool {
	if head == nil || head.Next == nil {
		return false
	}
    cur := head
    occur := make(map[*ListNode]int)
    for cur != nil {
        if _, ok := occur[cur]; ok {
            return true
        }
        occur[cur]++
        cur = cur.Next
    }
    return false
}

File: Data Structures & Algorithms/maximum-twin-sum-of-a-linked-list/submission-2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def pairSum(self, head: Optional[ListNode]) -> int:
        slow, fast = head, head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        prev, cur = None, slow
        while cur:
            nxt = cur.next
            cur.next = prev
            prev = cur
            cur = nxt

        res = 0
        l, r = head, prev
        while r:
            res = max(res, l.val + r.val)
            l, r = l.next, r.next

        return res